dora/core/common/src/main/java/alluxio/underfs/ObjectMultipartUploader.java

```diff
@@ -0,0 +1,178 @@
+  private final ObjectUnderFileSystem mUfs;
+  private final ListeningExecutorService mExecutor;
+
+  public String getContentHash() {
+    return mContentHash;
+  }
+
+  public ObjectMultipartUploader(String path, ObjectUnderFileSystem ufs,
+                                 ListeningExecutorService executor) {
+    mPath = path;
+    mUfs = ufs;
+    mExecutor = executor;
+  }
+
+  @Override
+  public void startUpload() throws IOException {
+    mUploadId = mUfs.initMultiPart(mPath, MultipartUfsOptions.defaultOption());
+  }
+
+  @Override
+  public ListenableFuture<Void> putPart(InputStream in, int partNumber, long partSize) throws IOException {
+    ListenableFuture<Void> f = mExecutor.submit(() -> {
+      try {
+        LOG.warn("put part for id {}, partnumber is {}.", mUploadId, partNumber);
+        String etag = mUfs.uploadPartWithStream(mPath, mUploadId, partNumber, partSize, in,
+            MultipartUfsOptions.defaultOption());
+        Pair<Integer, String> etagPair = new Pair<>(partNumber, etag);
+        mETags.put(partNumber, etagPair);
+        return null;
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    });
+    mUploadTasks.put(partNumber, new UploadTask(partNumber, partSize, f));
+    return f;
+  }
+
+  @Override
+  public ListenableFuture<Void> putPart(ByteBuffer b, int partNumber) throws IOException {
+    long partSize = b.remaining();
+    ListenableFuture<Void> f = mExecutor.submit(() -> {
+      try {
+        LOG.warn("put part for id {}, partnumber is {}.", mUploadId, partNumber);
+        InputStream in = new ByteBufferInputStream(b);
+        String etag = mUfs.uploadPartWithStream(mPath, mUploadId, partNumber, partSize, in,
+            MultipartUfsOptions.defaultOption());
+        Pair<Integer, String> etagPair = new Pair<>(partNumber, etag);
+        mETags.put(partNumber, etagPair);
+        return null;
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    });
+    mUploadTasks.put(partNumber, new UploadTask(partNumber, partSize, f));
+    return f;
+  }
+
+  @Override
+  public void complete() throws IOException {
+    // If there is no upload task, it means the file is empty.
+    // Just create an empty file and set up the permissions in this case.
+    if (mUploadTasks.isEmpty()) {
+      mUfs.createEmptyObject(mPath);
+      //TODO (wyy) get content hash
+      LOG.debug("Object store ufs multipart upload finished for {}", mPath);
+      return;
+    }
+
+    flush();
+    // complete
+    List<Pair<Integer, String>> etags = new ArrayList<>();
+    for (Map.Entry<Integer, UploadTask> entry : mUploadTasks.entrySet()) {
+      UploadTask uploadTask = entry.getValue();
+      if (uploadTask.finished()) {
+        Integer partNum = entry.getKey();
+        etags.add(mETags.get(partNum));
+      }
+    }
+    LOG.warn("complete mpu for id {}, mETags is {}.", mUploadId, etags);
```

Comment:
do we want to add a finally block to clean up the remaining parts in case MPU fails?

Reply:
yes, will do